ethexe/db/src/database.rs

```diff
@@ -106,132 +104,195 @@ impl Clone for Database {
+        for _ in 0..OffchainTransaction::BLOCK_HASHES_WINDOW_SIZE {
+            if block_hash == reference_block_hash {
+                return Ok(true);
+            }
+
+            let Some(block_header) = OnChainStorage::block_header(self, block_hash) else {
+                bail!(
+                    "Block with {block_hash} hash not found in the window. Possibly reorg happened"
+                );
+            };
+            block_hash = block_header.parent_hash;
+        }
+
+        Ok(false)
     }
 
-    fn set_block_end_state_is_valid(&self, block_hash: H256, is_valid: bool) {
-        log::trace!(target: LOG_TARGET, "For block {block_hash} set end state valid: {is_valid}");
-        let meta = self.block_small_meta(block_hash).unwrap_or_default();
-        self.set_block_small_meta(
-            block_hash,
-            BlockSmallMetaInfo {
-                block_end_state_is_valid: is_valid,
-                ..meta
-            },
-        );
+    fn with_small_data<R>(
+        &self,
+        block_hash: H256,
+        f: impl FnOnce(BlockSmallData) -> R,
+    ) -> Option<R> {
+        self.block_small_data(block_hash).map(f)
     }
 
-    fn block_is_empty(&self, block_hash: H256) -> Option<bool> {
-        self.block_small_meta(block_hash)
-            .and_then(|meta| meta.is_empty)
+    fn mutate_small_data(&self, block_hash: H256, f: impl FnOnce(&mut BlockSmallData)) {
+        let mut data = self.block_small_data(block_hash).unwrap_or_default();
+        f(&mut data);
+        self.set_block_small_data(block_hash, data);
     }
 
-    fn set_block_is_empty(&self, block_hash: H256, is_empty: bool) {
-        log::trace!(target: LOG_TARGET, "For block {block_hash} set is empty: {is_empty}");
-        let meta = self.block_small_meta(block_hash).unwrap_or_default();
-        self.set_block_small_meta(
-            block_hash,
-            BlockSmallMetaInfo {
-                is_empty: Some(is_empty),
-                ..meta
-            },
+    fn block_small_data(&self, block_hash: H256) -> Option<BlockSmallData> {
+        self.kv
+            .get(&KeyPrefix::BlockSmallData.two(self.router_address, block_hash))
+            .map(|data| {
+                BlockSmallData::decode(&mut data.as_slice())
+                    .expect("Failed to decode data into `BlockSmallMetaInfo`")
+            })
+    }
+
+    fn set_block_small_data(&self, block_hash: H256, meta: BlockSmallData) {
+        self.kv.put(
+            &KeyPrefix::BlockSmallData.two(self.router_address, block_hash),
+            meta.encode(),
         );
     }
+}
 
-    fn block_commitment_queue(&self, block_hash: H256) -> Option<VecDeque<H256>> {
-        self.block_small_meta(block_hash)
-            .and_then(|meta| meta.commitment_queue)
+#[cfg_attr(test, derive(serde::Serialize))]
+#[derive(Debug, Clone, Default, Encode, Decode, PartialEq, Eq)]
+struct BlockSmallData {
+    block_header: Option<BlockHeader>,
```

Comment:
why is it optional?

Reply:
if we have some hash that we must have header; event in sync process first of all we download all headers

Reply:
BlockSmallData is an internal structure which is just to optimise database implementation. For database BlockSmallData contains several in-depended fields each one can be set or not, so each field is Option, except bools which can be set by default.

Reply:
Discussed